### ISSUE TITLE
Bug: get "/object/stock" returns "null" for all userfield values - Root Cause: object_id for stock userfields in "userfield_value" table is "stock_id" instead of "object_id"

### ISSUE THREAD

**Issue by CRGer** [grocy/grocy#2380]
The object_id for **stock userfields** in "userfield_value" table is "stock_id" instead of "object_id". See screenshot of "userfield_values_resolves":
![image](https://github.com/grocy/grocy/assets/56159238/298bcb17-9771-448c-9e3b-c70c4649d921)

This results in at least the following issues while using the API:
- get "/object/stock" returns "null" for all userfield values
- get "/object/stock/{objectId} returns "null" for all userfield values
- get "/userfields/stock/{objectId} returns "null" for all userfield values (Note: For this route the objectid is a string, hence using stock_id is possible here. This reteurns the expected userfield values.)
- etc.

I could not find any issues in the frontend.

Sorry for not creating a pull request, but I'm...
- ...unsure if using the stock_id for stock userfield is intended.
- ...new to grocy and (currently) not so familiar with the code, so I want avoid to break things.

(I'm in the process of creating a scanner app that is tailored to my personal needs when using Grocy. For now, I have implemented a workaround in my app to make the API work as intended. The bug is therefore not too severe for me right now).

**Comment by berrnd** [comment]
> unsure if using the stock_id for stock userfield is intended

It is. Opening a Bug for something which you don't understand at first (which is of course ok) is the thing I like most, why not simply ask what's unclear for you?

Since you're new, I guess it's best to recap a little lesson on Grocy stock management basics + the technical relation behind:

- Products can be defined on [/products](https://demo-prerelease.grocy.info/products)
  - = `products` table
- A product can be in stock of course _n_ times (not only once), since that's how the reality works
- Such a product in stock is called a stock entry, those can be viewed and edited on [/stockentries](https://demo-prerelease.grocy.info/stockentries)
  - By nature, each and every stock entry can have a different price (and amount, due date, location and so on)
  - = `stock` table. There can be _n_ rows per stock entry, they're auto-splitted (e.g. on consuming a partial amount) and combined based on the fields which make a stock entry unique. Means, the column `id` can change at any time and is therefore not used as a reference on purpose. `stock_id` on the other was invented to clearly identify a unique stock entry, it stays the same for the whole lifecycle of the corresponding stock entry.
- Each and every stock change/transaction is recorded in a log called stock journal, visible on [/stockjournal](https://demo-prerelease.grocy.info/stockjournal)
  - = `stock_log` table
- [/stockoverview](https://demo-prerelease.grocy.info/stockoverview) is a "high level, kind of aggregated" overview what's in stock, means there is only one line per product
- At each and every place an amount can be entered, any product related unit can be used
  - The base is always the product's QU stock, there needs to exist a conversion (either a global or a product specific one) between that and the unit you want to use to make it selectable (since Grocy stores everything related to the products QU stock and it needs to know how much Pieces a Pack has, for example - it's impossible to magically guess that)

&nbsp;
> I could not find any issues in the frontend.

Which clearly leads to that there is no problem, since the web frontend uses the same API for kind of everything.

&nbsp;
If there is a concrete question on how to do stuff with the API, feel free to ask and I'm happy to assist, as always, instead of opening another Bug report here, thanks.

**Comment by CRGer** [comment]
Dear @berrnd,

First of all, thank you very much for your detailed reply. I don't think it came across clearly where I see the bug, possibly due to the structure of my post, so a few more points for clarification:

- I'm new, but I understand the functionality well, and also how stock entries work. I just haven't read the complete source code.
- The bug in my post does not refer to the fact that the object_id for user fields in the table "userfield_value" is "stock_id" instead of "object_id". This I see only as the lowest level of the reason (not bug!) that the API does not work as expected in this case. Between the fact (database structure for stock entries) and the "bug" there are of course more levels, such as the technical functioning of the API on the server side. To show again where I see the bug, I described it below again (better structured and with example).

---

**Demo Instance**: https://test-1oqawcd8i1ki25lvbll7z5oi.demo-prerelease.grocy.info/

**configured userfield**:
![image](https://github.com/grocy/grocy/assets/56159238/7984963b-676d-4ff5-a483-c2fbccda61af)

### Situation

**"/stockentries" shows value of testuserfield:**

![image](https://github.com/grocy/grocy/assets/56159238/dccd6354-bf78-4c0d-9805-9e9da5fe4f14)

**Calling "api/objects/stock" returns "null" for all testuserfield values**:

Request:
`curl -X 'GET' \
  'https://test-1oqawcd8i1ki25lvbll7z5oi.demo-prerelease.grocy.info/api/objects/stock' \
  -H 'accept: application/json' \
  -H 'GROCY-API-KEY: C04LL7V9iI5Y1jbBadoMxiXkWoEBfa7FBbrbEJauMLSnHcP9N5'`

Response:
```
[
  {
    "id": 1,
    "product_id": 1,
    "amount": 1,
    "best_before_date": "2023-11-15",
    "purchased_date": "2023-11-14",
    "stock_id": "655359ddb3acf",
    "price": null,
    "open": 0,
    "opened_date": null,
    "row_created_timestamp": "2023-11-14 12:28:29",
    "location_id": 2,
    "shopping_location_id": null,
    "note": null,
    "userfields": {
      "testuserfield": null
    }
  },
  {
    "id": 2,
    "product_id": 1,
    "amount": 2,
    "best_before_date": "2023-11-16",
    "purchased_date": "2023-11-14",
    "stock_id": "655359ede3821",
    "price": null,
    "open": 0,
    "opened_date": null,
    "row_created_timestamp": "2023-11-14 12:28:45",
    "location_id": 2,
    "shopping_location_id": null,
    "note": null,
    "userfields": {
      "testuserfield": null
    }
  },
  {
    "id": 3,
    "product_id": 2,
    "amount": 2,
    "best_before_date": "2023-11-17",
    "purchased_date": "2023-11-14",
    "stock_id": "65535a02bd317",
    "price": null,
    "open": 0,
    "opened_date": null,
    "row_created_timestamp": "2023-11-14 12:29:06",
    "location_id": 2,
    "shopping_location_id": null,
    "note": null,
    "userfields": {
      "testuserfield": null
    }
  },
  {
    "id": 4,
    "product_id": 2,
    "amount": 3,
    "best_before_date": "2023-11-22",
    "purchased_date": "2023-11-14",
    "stock_id": "65535a0b7a960",
    "price": null,
    "open": 0,
    "opened_date": null,
    "row_created_timestamp": "2023-11-14 12:29:15",
    "location_id": 2,
    "shopping_location_id": null,
    "note": null,
    "userfields": {
      "testuserfield": null
    }
  }
]
```

### Expected Behavour

**Calling "api/objects/stock" return the actual value for testuserfield of all stock entries.**

---
I hope the explanation has now clarified things. Thank you for your time and thank you for all the work you have put into Grocy!

**Comment by berrnd** [comment]
What you want and expect is just not how it's built. A Bug is something that doesn't work, not something that is not (yet) done or built differently than you personally expect it to work. I'm always happy to answer and assist on any question, most of the time within hours for each and every request in the last more than 6 years Grocy. But I'm allergic to reporting each and everything that comes up on a first sight as a Bug instead of just asking why it is how it is or how it's supposed to work (since it works, the proof is always the web frontend, I'm dogfooding Grocy's ~great~ bad API myself, since Grocy exists), that was all regarding that - back to topic:

Technically the entity handling and Userfields are all built in a kind of low level and generic way. So of course Userfields for stock entries are an exception because of using a different id as the reference ([example code ref](https://github.com/grocy/grocy/blob/dc05476d09e40f6e4007e39cf31ac6f66c24eb30/controllers/GenericEntityApiController.php#L236) regarding Userfield handling for all [generic entity interaction endpoints](https://test-1oqawcd8i1ki25lvbll7z5oi.demo-prerelease.grocy.info/api#/Generic%20entity%20interactions)). It's nowhere stated the every detail works across every entity.

The official Userfield endpoints are [` GET/PUT /userfields/{entity}/{objectId}`](https://demo-prerelease.grocy.info/api#/Generic%20entity%20interactions/get_userfields__entity___objectId_).

And that also works just fine for your example - picking a random stock entry: https://test-1oqawcd8i1ki25lvbll7z5oi.demo-prerelease.grocy.info/api/userfields/stock/655359ddb3acf

Sometimes two (or even more) API calls are needed reach the desired goal. But since this is not a cloud service which charges money per API call, I don't really see the point what's the problem on that in general.

So in conclusion: Of course what you expected can be added / improved, but it's just not a Bug, since it works how it's defined to work.

**Comment by berrnd** [comment]
I've created #2381 to track the corresponding Feature Request ([as always](https://github.com/grocy/grocy#roadmap) without any promises on when it's done, reporting wishes as a Bug lowered my desire to work on it drastically so far, but let's see).
